### ISSUE TITLE
Reordering relation by dragging in UI doesn't persist after save

### ISSUE THREAD

**Issue by ericbf** [strapi/strapi#7039]
**Describe the bug**
Reordering relation by dragging in UI doesn't persist after save.

**Steps to reproduce the behavior**

1. Create two content types, one with a one to many relation to the other
1. Add an item to the "parent" type and some items to the "children" type
1. In the parent type, add some children to the relation and save
1. Reopen the parent item, change the order of the children in the relation and save
1. Reopen the parent and look at the order of the children

**Expected behavior**
The new order you just saved should be reflected

**Actual behavior**
The new order is not saved and the old order is reflected.

**Screenshots**
Start state:
![Screen Shot 2020-07-16 at 10 55 12 AM](https://user-images.githubusercontent.com/2483476/87686911-1793da00-c753-11ea-97c8-6d1fb1ccb723.png)
After reordering:
![Screen Shot 2020-07-16 at 10 55 21 AM](https://user-images.githubusercontent.com/2483476/87686930-1e225180-c753-11ea-9888-dd3d5fa8d704.png)
After saving and reopening:
![Screen Shot 2020-07-16 at 10 55 32 AM](https://user-images.githubusercontent.com/2483476/87686942-237f9c00-c753-11ea-8262-d10b3ac54d7e.png)

**System**

- Node.js version: v12.18.2
- NPM version: 6.14.5
- Strapi version: v3.0.5
- Database: sqlite3
- Operating system: MacOS Catalina, Version 10.15.5

**Comment by derrickmehaffy** [comment]
Currently relations don't have any means to save their order within the database, (UI supports it, but the data isn't saved anywhere).

See the following FAQ about this: https://strapi.io/documentation/v3.x/getting-started/troubleshooting.html#relations-aren-t-maintaining-their-sort-order

The current implementation is the intended one, so for now I'm going to close this. If/when the Strapi team decide if the storage is a viable option (there is a lot of backend requirements for this) then we can take another look at it.

**Comment by ericbf** [comment]
@derrickmehaffy is there a way to hook into the UI action in order to properly process the action that is supported in the UI to be reflected in the backing store?

**Comment by derrickmehaffy** [comment]
I believe you would need to custom either the Content-Manager controller or service

Controller: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-content-manager/controllers/ContentManager.js
Service: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-content-manager/services/ContentManager.js

And possibly override the admin panel code within that same plugin to actually send the order, but as I mentioned before currently there is nothing in the backend to actually save this. Components have a field added called `order` to save this data (and likewise a filter is preapplied) but with relations as the FAQ describes can be much more complex especially when you start dealing with manyToMany and manyWay relations that can be attached to multiple models and have many different sorting lists.

My suggestion to you would be to apply a sorting parameter to the relations based on another field or using the [customize the data response](https://strapi.io/documentation/v3.x/guides/custom-data-response.html) to do so based on a specific field.

Another key problem with sorting of relations is the possible amount, if every relation value had an order field and you moved one of them up or down in the sort, you could very easily cause a cascading performance issue updating all the other values with the new order. A failure here could cause massive corruption if the query fails part way through on the database side.

**Comment by ericbf** [comment]
I'm trying to solve this by adding a hidden field to the component/content-type that contains the relation, but I can't seem to pin down which files I'd need to edit. Can you help me out? When the ContentManager controller sends the payload to the ContentManager service, the order that the user selected is *already* sent, so there are no changes necessary to the admin UI.

**Comment by ericbf** [comment]
Basically I want to intercept the edit/create calls to persist the order to this hidden field, and intercept the fetch calls to recall the order from the field

**Comment by ericbf** [comment]
I also need to intercept some content-type-builder call to actually add that hidden field, prevent it from being shown, and etc.

**Comment by derrickmehaffy** [comment]
@ericbf currently there is nothing that sends this payload, it's only there in the UI but doesn't actually send anything currently. I'll ping our frontend guys @soupette and @HichamELBSI but right now I'm not sure there is much help we can provide as we haven't really built the feature ourselves.

**Comment by ericbf** [comment]
Inside the ContentManager.js edit function, if I log the data, I see the relation as an array of IDs, whose order reflects the latest I selected in the UI. I'm basically looking for how to create a hidden field where I can store a custom string. I think I can take care of all the rest based on what I have.

**Comment by derrickmehaffy** [comment]
You can use the security validations: https://strapi.io/documentation/v3.x/concepts/models.html#validations

Such as setting private to true

**Comment by ericbf** [comment]
Excellent! Nextly, is there a way to create this field automatically whenever a relation field is added to an element? That would be awesome, but if not, I could just do it manually. There's not too too many relations in my project.

**Comment by derrickmehaffy** [comment]
No not currently :(

**Comment by ericbf** [comment]
Okay, that's alright. Let me run with this and see how it works out. Thanks!

**Comment by ericbf** [comment]
Got it working for our use case with the above advices. My solution was three prong: here's what I did:

1. I added a string field alongside every relation whose order should be maintained, named with the same name as the relation, but prefixed with `order_`. This field must **not** be private, and should be hidden from the "view" so as it is not shown in the actual content entry form.

2. Add a custom implementation of `content-manager/service/ContentManager.js`. This records the order of any relation that appears alongside an `order_*` prop. I wasn't super concerned about optimizing it, so I just use `JSON.stringify`, as the relation comes through as an ID array, conveniently. The implementation also recalls the saved order for the `content-manager` UI, such that the order is persistent.
```js
"use strict"

const _ = require("lodash")

/**
 * A set of functions called "actions" for `ContentManager`
 */
module.exports = {
	async fetchAll(model, query) {
		const { query: request, populate, ...filters } = query

		const queryFilter = !_.isEmpty(request)
			? {
					...filters, // Filters is an object containing the limit/sort and start
					...request
			  }
			: filters

		const found = await strapi.entityService.find(
			{
				params: queryFilter,
				populate
			},
			{ model }
		)

		recallOrder(found)

		return found
	},

	async fetch(model, id, config = {}) {
		const { query = {}, populate } = config

		const found = await strapi.entityService.findOne(
			{
				params: { ...query, id },
				populate
			},
			{ model }
		)

		recallOrder(found)

		return found
	},

	count(model, query) {
		return strapi.entityService.count({ params: query }, { model })
	},

	create({ data, files }, { model } = {}) {
		return strapi.entityService.create({ data, files }, { model })
	},

	edit(params, { data, files }, { model } = {}) {
		saveOrder(data)

		console.log(data)

		return strapi.entityService.update({ params, data, files }, { model })
	},

	delete(model, query) {
		return strapi.entityService.delete({ params: query }, { model })
	},

	deleteMany(model, ids, query) {
		const { primaryKey } = strapi.query(model)

		return strapi.entityService.delete(
			{
				params: {
					_limit: 100,
					...query,
					_where: _.concat({ [`${primaryKey}_in`]: ids }, query._where || {})
				}
			},
			{ model }
		)
	},

	search(model, query, params) {
		return strapi.entityService.search({ params: { ...query, ...params } }, { model })
	},

	countSearch(model, query) {
		return strapi.entityService.countSearch({ params: query }, { model })
	}
}

/**
 * Persist the order of relations in the passed data.
 * @param {any} data
 */
function saveOrder(data) {
	manageOrders(data, (data, orderKey, key) => {
		if (!Array.isArray(data[key]) || data[key].some((o) => typeof o !== "number")) {
			return
		}

		data[orderKey] = JSON.stringify(data[key])
	})
}

/**
 * Recall the order of relations in the passed data.
 * @param {any} data
 */
function recallOrder(data) {
	manageOrders(data, (data, orderKey, key) => {
		try {
			/** @type {number[] | {}} */
			const order = JSON.parse(data[orderKey])

			/** @type {number[]} */
			const currentOrder = data[key].map((i) => i.id)
			/** @type {{ [key: number]: any }} */
			const currentItems = data[key].reduce((trans, next) => {
				trans[next.id] = next

				return trans
			}, {})

			if (!Array.isArray(order) || order.some((o) => typeof o !== "number")) {
				return
			}

			const newOrder = order.slice(0)

			// Remove all elements from the new order that have been removed from the relation
			for (let i = newOrder.length - 1; i >= 0; i -= 1) {
				if (!currentOrder.includes(newOrder[i])) {
					newOrder.splice(i, 1)
				}
			}

			// Append all elements to the new order that have been added in the relation
			for (const item of currentOrder) {
				if (!newOrder.includes(item)) {
					newOrder.push(item)
				}
			}

			data[key].length = 0
			data[key].push(...newOrder.map((i) => currentItems[i]))
		} catch {}
	})
}

/**
 * Apply the order of relations in the passed data.
 * @param {any} data
 * @param {(data: any, orderKey: string, key: string) => void} callback
 */
function manageOrders(data, callback) {
	if (typeof data !== "object" || data == null) {
		return
	}

	const orderRegex = /^order_/
	const orderKeys = Object.keys(data).filter((k) => orderRegex.test(k))

	for (const orderKey of orderKeys) {
		const key = orderKey.replace(orderRegex, "")

		if (key in data && Array.isArray(data[key])) {
			callback(data, orderKey, key)
		}
	}

	for (const key of Object.keys(data)) {
		const isOrdered = orderKeys.some((order) => {
			return key === order || key === order.replace(orderRegex, "")
		})

		if (!isOrdered) {
			manageOrders(data[key], callback)
		}
	}
}
```

3. Add a middleware to recall the order on API calls. This doesn't bother with updating through API, as my use case didn't call for it, but you would probably add something similar to this for that use case.
```js
module.exports = (strapi) => {
	return {
		initialize() {
			strapi.app.use(async (ctx, next) => {
				await next()

				if (!ctx.request.url.includes("/content-manager/")) {
					console.log("do something!")

					recallOrder(ctx.body)
				}
			})
		}
	}
}

/**
 * Recall the order of relations in the passed data.
 * @param {any} data
 */
function recallOrder(data) {
	manageOrders(data, (data, orderKey, key) => {
		try {
			/** @type {number[] | {}} */
			const order = JSON.parse(data[orderKey])

			/** @type {number[]} */
			const currentOrder = data[key].map((i) => i.id)
			/** @type {{ [key: number]: any }} */
			const currentItems = data[key].reduce((trans, next) => {
				trans[next.id] = next

				return trans
			}, {})

			if (!Array.isArray(order) || order.some((o) => typeof o !== "number")) {
				return
			}

			const newOrder = order.slice(0)

			// Remove all elements from the new order that have been removed from the relation
			for (let i = newOrder.length - 1; i >= 0; i -= 1) {
				if (!currentOrder.includes(newOrder[i])) {
					newOrder.splice(i, 1)
				}
			}

			// Append all elements to the new order that have been added in the relation
			for (const item of currentOrder) {
				if (!newOrder.includes(item)) {
					newOrder.push(item)
				}
			}

			data[key].length = 0
			data[key].push(...newOrder.map((i) => currentItems[i]))

			delete data[orderKey]
		} catch {}
	})
}

/**
 * Apply the order of relations in the passed data.
 * @param {any} data
 * @param {(data: any, orderKey: string, key: string) => void} callback
 */
function manageOrders(data, callback) {
	if (typeof data !== "object" || data == null) {
		return
	}

	const orderRegex = /^order_/
	const orderKeys = Object.keys(data).filter((k) => orderRegex.test(k))

	for (const orderKey of orderKeys) {
		const key = orderKey.replace(orderRegex, "")

		if (key in data && Array.isArray(data[key])) {
			callback(data, orderKey, key)
		}
	}

	for (const key of Object.keys(data)) {
		const isOrdered = orderKeys.some((order) => {
			return key === order || key === order.replace(orderRegex, "")
		})

		if (!isOrdered) {
			manageOrders(data[key], callback)
		}
	}
}
```

As I said above, this isn't aimed at being optimized. It currently is O(n^2) from my quick estimation. But our relations that need order persisted are not going about 20ish relations, so it's not a big deal.

My suggestion to the strapi team is to implement something similar (but more optimized), but not as default for relations—perhaps an option under advanced for relations that it should be persistent order or not, defaulting to false?

**Comment by derrickmehaffy** [comment]
Thank you @ericbf

**Comment by roelbeerens** [comment]
Hiya! Just chippin in here, because I had a similar issue at my project. What I ended up doing was firing a raw database query in the beforeUpdate of a Content-Type. Let's say we have a type for pages and that has a relation with blocks. Reordering and updating them in MySQL doesn't work, so what I did in the beforeUpdate of `page` was this:

```
await strapi.connections.default.raw(`DELETE FROM pages__blocks WHERE page_id = ${data.id};`);
```

This removes the current set blocks relations for the page. However, because in our data object, we have a new set of block relations with the newly set order, it creates those relations on update.

It's a simple solution (I know it's not super efficient) but it does the job for us. Maybe this helps to shine some light on people's project having the same issue.
